### PR TITLE
Fix motion direction in slope for CharacterBody3D

### LINKED ISSUE #52958: `CharacterBody3D` causes slipping when going up the slope with slanting to it
### Godot version

4.0.dev

### System information

MacOS 11.4

### Issue description

If you try to go up the slope **with slanting to the slope**, there will be a gap with the target point, which will result in a slip.

### Steps to reproduce

https://user-images.githubusercontent.com/61938263/134451477-074a2b1b-265f-42dd-84a7-85ff86c99e8f.mov

This slipping is very uncomfortable for the player because it causes a gap between the player's viewed target and the input.

スクリーンショット 2021-09-23 12 22 30

This is one of the parts where the 2D and 3D calculations differ significantly. For reference, before #51904 was merged, there was a way to avoid this slip on the GDScript.

```gd
# Apply GRAVITY
_state_velocity = Vector3(_state_velocity.x, _state_velocity.y + (-GRAVITY * delta), _state_velocity.z)
if self.is_on_floor():
	var length = Vector3(_state_velocity.x, 0, _state_velocity.z).length()
	_state_velocity = _state_velocity.normalized() * length
else:
	_state_velocity = Vector3(_state_velocity.x, _state_velocity.y + (-GRAVITY * delta), _state_velocity.z)
```

https://user-images.githubusercontent.com/61938263/134451711-83eeb43b-d0c0-4322-87b4-1b65f06a6afe.mov

However, the above code is based on the default behavior of "increasing movement speed on slopes" before the merge of #51904.

I think the fundamental problem is that the constant velocity is based on the vector that slides down by #51904. I think we should change the algorithm of the sliding down, because the sliding down by #51904 already causes a gap with the target.

### Minimal reproduction project

#50732

### PR DESCRIPTION
**Pull request by pouleyKetchoupp**
Fixes #52958

List of changes:
-More accurate sliding in slopes to keep input direction correct
-More accurate constant speed calculation
-Reset gravity accumulation was done in two places
-Renamed linear_velocity to motion_velocity for clarity
-General code cleaning and simplifications

### CODE REVIEW

File: scene/3d/physics_body_3d.cpp [hunk 1]
```diff
@@ -1305,22 +1319,28 @@ void CharacterBody3D::_move_and_slide_grounded(double p_delta, bool p_was_on_flo
...
-			Vector3 motion_slide_norm = motion.slide(prev_floor_normal).normalized();
+			// Slide using the intersection between the motion plane and the floor plane,
+			// in order to keep the direction intact.
+			Vector3 motion_slide_norm = up_direction.cross(motion).cross(prev_floor_normal);
+			motion_slide_norm.normalize();
+
 			motion = motion_slide_norm * (motion_slide_up.length());
 			collided = true;
+		} else {
+			// No collision.
```
Comment (fabriceci) [inline scene/3d/physics_body_3d.cpp:1330]: If there is no collision, `last_motion` should be equals to `motion`. Seems to be empty here, you can remove this `else` condition to let the line `1343` do the job.

File: scene/3d/physics_body_3d.cpp [hunk 2]
```diff
@@ -1367,13 +1387,19 @@ void CharacterBody3D::_move_and_slide_free(double p_delta) {
 			if (motion.dot(linear_velocity) <= 0.0) {
 				motion = Vector3();
 			}
+		} else {
+			// No collision.
```
Comment (fabriceci) [inline scene/3d/physics_body_3d.cpp:1391]: same here

**Review by fabriceci — APPROVED** [review]
Looks good, just check the 2 comments.

**Comment by pouleyKetchoupp** [comment]
Updated with fixes for `last_motion` and also renamed `linear_velocity` to `motion_velocity` as in #53149.

File: doc/classes/CharacterBody3D.xml [hunk 3]
```diff
@@ -147,17 +147,17 @@
 			As long as the snapping vector is in contact with the ground and the body moves against `up_direction`, the body will remain attached to the surface. Snapping is not applied if the body moves along `up_direction`, so it will be able to detach from the ground when jumping.
 		</member>
 		<member name="floor_stop_on_slope" type="bool" setter="set_floor_stop_on_slope_enabled" getter="is_floor_stop_on_slope_enabled" default="true">
-			If [code]true[/code], the body will not slide on slopes when you include gravity in [code]linear_velocity[/code] when calling [method move_and_slide] and the body is standing still.
-		</member>
-		<member name="linear_velocity" type="Vector3" setter="set_linear_velocity" getter="get_linear_velocity" default="Vector3(0, 0, 0)">
-			Current velocity vector (typically meters per second), used and modified during calls to [method move_and_slide].
+			If [code]true[/code], the body will not slide on slopes when calling [method move_and_slide] when the body is standing still.
```
Comment (fabriceci) [inline doc/classes/CharacterBody3D.xml:150]: We need the decide that (conversation in the other PR) before the merge.

Comment (pouleyKetchoupp) [reply doc/classes/CharacterBody3D.xml:150]: Just updated to explain how it slides when stop on slope is disabled.

**Pull request merged**